cabal-install/src/Distribution/Client/ProjectPlanning.hs

```diff
@@ -2514,11 +2514,11 @@ nubComponentTargets =
     -- If we're building the whole component then that the only target all we
     -- need, otherwise we can have several targets within the component.
     wholeComponentOverrides :: [(ComponentTarget,  a )]
-                            -> [(ComponentTarget, [a])]
+                            -> [(ComponentTarget, NonEmpty a)]
     wholeComponentOverrides ts =
       case [ t | (t@(ComponentTarget _ WholeComponent), _) <- ts ] of
-        (t:_) -> [ (t, map snd ts) ]
-        []    -> [ (t,[x]) | (t,x) <- ts ]
+        (t:_) -> [ (t, NE.fromList (map snd ts)) ]
```

Comment:
Only unsafe function usage, should the pre-condition be documented somehow?

Reply:
yeah, just an inline comment should be fine here

Reply:
If it isn't much trouble I'd follow phadej's advice here. Dumping "NonEmpty.fromList: empty list" on the user isn't too nice

Reply:
I came up with the following solution:

```haskell
    wholeComponentOverrides :: [(ComponentTarget,  a )]
                            -> [(ComponentTarget, NonEmpty a)]
    wholeComponentOverrides ts =
      case [ ta | ta@(ComponentTarget _ WholeComponent, _) <- ts ] of
        ((t, x):_) -> 
                let
                    -- Delete tuple (t, x) from original list to avoid duplicates.
                    -- Use 'deleteBy', to avoid additional Class constraint on 'nubComponentTargets'.
                    ts' = deleteBy (\(t1, _) (t2, _) -> t1 == t2) (t, x) ts
                in
                    [ (t, x :| map snd ts') ]
        []    -> [ (t, x :| []) | (t,x) <- ts ]
```

Longer, but no unsafe function!